src/couch_replicator/src/couch_replicator_auth_session.erl

```diff
@@ -0,0 +1,545 @@
+% License for the specific language governing permissions and limitations under
+% the License.
+
+
+% This is the replicator session auth plugin. It implements session based
+% authentication for the replicator. The only public API are the functions from
+% the couch_replicator_auth behaviour. Most of the logic and state is in the
+% gen_server. An instance of a gen_server could be spawned for the source and
+% target endpoints of each replication jobs.
+%
+% The workflow is roughly this:
+%
+%  * On initialization, try to get a cookie in `refresh/1` If an error occurs,
+%    the crash. If `_session` endpoint fails with a 404 (not found), return
+%    `ignore` assuming session authentication is not support or we simply hit a
+%    non-CouchDb server.
+%
+%  * Before each request, auth framework calls `update_headers` API function.
+%    Before updating the headers and returning, check if need to refresh again.
+%    The check looks `next_refresh` time. If that time is set (not `infinity`)
+%    and just expired, then obtain a new cookie, then update headers and
+%    return.
+%
+%  * After each request, auth framework calls `handle_response` function. If
+%    request was successful check if a new cookie was sent by the server in the
+%    `Set-Cookie` header. If it was then then that becomes the current cookie.
+%
+%  * If last request has an auth failure, check if request used a stale cookie
+%    In this case nothing is done, and the client is told to retry. Next time
+%    it updates its headers befor the request it should pick up the latest
+%    cookie.
+%
+%  * If last request failed and cookie was the latest known cookie, schedule a
+%    refresh and tell client to retry. However, if the cookie was just updated,
+%    tell the client to continue such that it will handle the auth failure on
+%    its own via a set of retries with exponential backoffs. This is it to
+%    ensure if something goes wrong and one of the endpoints issues invalid
+%    cookies, replicator won't be stuck in a busy loop refreshing them.
+
+
+-module(couch_replicator_auth_session).
+
+
+-behaviour(couch_replicator_auth).
+-behaviour(gen_server).
+
+
+-export([
+    initialize/1,
+    update_headers/2,
+    handle_response/4,
+    cleanup/1
+]).
+
+-export([
+    init/1,
+    terminate/2,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2,
+    code_change/3,
+    format_status/2
+]).
+
+
+-include_lib("ibrowse/include/ibrowse.hrl").
+-include_lib("couch_replicator/include/couch_replicator_api_wrap.hrl").
+
+
+-type headers() :: [{string(), string()}].
+-type code() :: non_neg_integer().
+-type creds() :: {string() | undefined, string() | undefined}.
+
+
+% Behavior API callbacks
+
+
+-spec initialize(#httpdb{}) -> {ok, #httpdb{}, term()} | ignore.
+initialize(#httpdb{} = HttpDb) ->
+    case remove_creds(HttpDb) of
+        {ok, User, Pass, HttpDb1} ->
+            case gen_server:start_link(?MODULE, [User, Pass, HttpDb1], []) of
```

Comment:
talk to @davisp, there's a defined pattern to follow for gen_servers that might fail at startup.

Reply:
You want to use proc_lib:start_link, proc_lib:init_ack, and gen_server:enter_loop. Here's an example from config_listener_mon.erl:

https://github.com/apache/couchdb-config/blob/1.0.2/src/config_listener_mon.erl#L40-L57

Basically, proc_lib:start_link/proc_lib:init_ack allow you to do a synchronous process startup and the new process can return anything it wants. If your gen_server state creation thing succeeds you can then just call gen_server:enter_loop to become a gen_server. Or else you just init_ack any error and return and the process dies.

Reply:
But what's the downside of letting start_link fail with an error vs doing the manual proc_lib trick?

Looking at docs at http://erlang.org/doc/man/gen_server.html#start_link-3 `{error, Reason}` seems like it would do what we want?

```
If the gen_server process is successfully created and initialized, the function returns {ok,Pid}, where Pid is the pid of the gen_server process. If a process with the specified ServerName exists already, the function returns {error,{already_started,Pid}}, where Pid is the pid of that process.

If Module:init/1 fails with Reason, the function returns {error,Reason}. If Module:init/1 returns {stop,Reason} or ignore, the process is terminated and the function returns {error,Reason} or ignore, respectively.
```

Reply:
I moved initialization, where a lot of errors happen, out of the gen_server's `init/1` to the caller's process in `initialize/1. This way errors look a bit cleaner.